velox/exec/Spill.h

```diff
@@ -429,26 +429,18 @@ class FileSpillBatchStream : public BatchStream {
 /// start bit offset is used to calculate the partition number of spill data. It
 /// is required for the recursive spilling handling as we advance the start bit
 /// offset when we go to the next level of recursive spilling.
-struct SpillPartitionId {
-  SpillPartitionId()
-      : SpillPartitionId(
-            std::numeric_limits<uint8_t>::max(),
-            std::numeric_limits<int32_t>::max()) {}
-
-  SpillPartitionId(uint8_t _partitionBitOffset, int32_t _partitionNumber)
-      : partitionBitOffset(_partitionBitOffset),
-        partitionNumber(_partitionNumber) {}
-
-  uint8_t partitionBitOffset = 0;
-  int32_t partitionNumber = 0;
-
-  bool valid() const {
-    return !(*this == SpillPartitionId());
-  }
+///
+/// NOTE: multiple shards created from the same SpillPartition by split()
+/// will share the same id.
+class SpillPartitionId {
+ public:
+  SpillPartitionId(uint8_t partitionBitOffset, int32_t partitionNumber)
+      : partitionBitOffset_(partitionBitOffset),
+        partitionNumber_(partitionNumber) {}
 
   bool operator==(const SpillPartitionId& other) const {
-    return std::tie(partitionBitOffset, partitionNumber) ==
-        std::tie(other.partitionBitOffset, other.partitionNumber);
+    return std::tie(partitionBitOffset_, partitionNumber_) ==
```

Comment:
Say partition p1 and p2 are from layer 1, and they are both spilled, then their children might have the same partitionNumber, so children from different parents can have the same `SpillPartitionId`, is the understanding correct? cc @xiaoxmeng @mbasmanova If the understanding is correct, then the check `VELOX_CHECK_EQ(spillPartitionSets_.count(id), 0);` in `HashJoinBridge::setHashTable` might fail in such case? Please correct me if I'm wrong with this...

Reply:
Ignore this... the `operator<` in `SpillPartitionId` ensures that the spill partition with higher `partitionBitOffset_` is always processed first.